src/openslide-vendor-zeiss.c

```diff
@@ -360,127 +360,130 @@ static void bgr48_to_argb32(uint8_t *src, size_t src_len, uint32_t *dst) {
+    const struct zisraw_zstd_payload_hdr *hdr =
+      (const struct zisraw_zstd_payload_hdr *) src;
+    // check that we can read the size field, then use it to check that we
+    // can read the whole header
+    if (len < (int64_t) sizeof(hdr->size) || len < hdr->size) {
+      g_set_error(err, OPENSLIDE_ERROR, OPENSLIDE_ERROR_FAILED,
+                  "Image data length %"PRId64" too small for zstd header", len);
+      return false;
+    }
+    switch (hdr->size) {
+    case 1:
+      break;
+    case 3:
+      if (hdr->chunk_type != 1) {
+        g_set_error(err, OPENSLIDE_ERROR, OPENSLIDE_ERROR_FAILED,
+                    "Unexpected zstd chunk type: %d", hdr->chunk_type);
+        return false;
+      }
+      do_hilo = hdr->is_hi_low_pack & 1;
+      break;
+    default:
+      g_set_error(err, OPENSLIDE_ERROR, OPENSLIDE_ERROR_FAILED,
+                  "Unexpected zstd header length: %u", hdr->size);
+      return false;
+    }
+    src += hdr->size;
+    len -= hdr->size;
+    // fall through
+  case COMP_ZSTD0:
+    // decompress
+    decompressed_data =
+      _openslide_zstd_decompress_buffer(src, len, pixel_bytes, err);
+    if (!decompressed_data) {
+      g_prefix_error(err, "Decompressing pixel data: ");
+      return false;
+    }
+    src = decompressed_data;
     break;
   default:
     g_assert_not_reached();
   }
 
-  g_autofree char *inbuf = g_malloc(len);
-  if (!freadn_to_buf(f, pos, inbuf, len, err)) {
-    g_prefix_error(err, "Couldn't read czi zstd compressed payload: ");
-    return false;
-  }
-
-  struct zisraw_zstd_payload_hdr hdr;
-  hdr.size = 0; /* zstd0 compression doesn't add prefix header */
-  hdr.is_hi_low_pack = 0;
-  if (compression == COMP_ZSTD1 && !parse_zstd_payload_hdr(&hdr, inbuf)) {
-    g_set_error(err, OPENSLIDE_ERROR, OPENSLIDE_ERROR_FAILED,
-                "Couldn't parse czi zstd1 compressed payload header");
-    return false;
-  }
-
-  size_t img_len = w * h * pixel_bytes;
-  g_autofree uint8_t *img = _openslide_zstd_decompress_buffer(
-      inbuf + hdr.size, len - hdr.size, img_len, err);
-  if (img == NULL) {
-    return false;
-  }
-
-  /* restore image modified by czi zstd1 high low byte pack trick */
-  if (hdr.is_hi_low_pack == 1) {
-    g_autofree uint8_t *tmp = g_malloc(img_len);
-    _openslide_restore_czi_zstd1(img, img_len, tmp);
-    convert(tmp, img_len, dst);
-  } else {
-    convert(img, img_len, dst);
+  // zstd1 compression has an option to pack less significant bytes of 16-
+  // bit pixels in the first half of the image array, and more significant
+  // bytes in the second half of the image array.  Undo this.
+  g_autofree uint8_t *unhilo_data = NULL;
+  if (do_hilo) {
+    if (pixel_bytes % 2) {
+      g_set_error(err, OPENSLIDE_ERROR, OPENSLIDE_ERROR_FAILED,
+                  "Can't perform HiLo unpacking with an odd number of bytes "
+                  "%"PRId64, pixel_bytes);
+      return false;
+    }
+    int64_t half_bytes = pixel_bytes / 2;
```

Comment:
Later when add SIMD, the undo HiLo pack will be moved to an individual function, which will be dispatched to pure C or SIMD, based on detected CPU features.